src/EditorFeatures/Core/Implementation/Outlining/OutliningTaggerProvider.cs

```diff
@@ -74,17 +97,75 @@ protected override ITaggerEventSource CreateEventSource(ITextView textViewOpt, I
                 TaggerEventSources.OnWorkspaceRegistrationChanged(subjectBuffer, TaggerDelay.OnIdle));
         }
 
-        protected override ITagProducer<IOutliningRegionTag> CreateTagProducer()
+        public override async Task ProduceTagsAsync(TaggerContext<IOutliningRegionTag> context, DocumentSnapshotSpan documentSnapshotSpan, int? caretPosition)
         {
-            return CreateConcreteTagProducer();
+            try
+            {
+                var cancellationToken = context.CancellationToken;
+                using (Logger.LogBlock(FunctionId.Tagger_Outlining_TagProducer_ProduceTags, cancellationToken))
+                {
+                    var document = documentSnapshotSpan.Document;
+                    var snapshotSpan = documentSnapshotSpan.SnapshotSpan;
+                    var snapshot = snapshotSpan.Snapshot;
+
+                    if (document != null)
+                    {
+                        var outliningService = document.Project.LanguageServices.GetService<IOutliningService>();
+                        if (outliningService != null)
+                        {
+                            // TODO: change this to shared pool once Esent branch RI
```

Comment:
I think we can remove the "TODO", I was actually trying to change API to rather than return List, It is passed into the API but @CyrusNajmabadi wanted prettier API rather than save allocations.